### ISSUE TITLE
iOS: UI will be blocked when show Alert while closing Modal

### ISSUE THREAD

**Issue by nihgwu** [facebook/react-native#10471]
When show Alert while closing Modal, the Alert dialog will disappear and the Modal will block the UI entirely even after reload, only on iOS.

``` js
'use strict';

import React, { Component } from 'react';

import {
  StyleSheet,
  View,
  Text,
  Modal,
  Alert,
} from 'react-native';

class demo extends Component {
  state = {
    showModal: false,
  }

  onShowModal = () => {
    this.setState({ showModal: true });
  }
  onCloseModal1 = () => {
    this.setState({ showModal: false }, () => {
      Alert.alert('Alert', 'UI will be blocked by the modal');
    });
  }
  onCloseModal2 = () => {
    this.setState({ showModal: false }, () => {
      setTimeout(() => {
        Alert.alert('Alert', 'Alert won\'t show');
      }, 200);
    });
  }
  onCloseModal3 = () => {
    this.setState({ showModal: false }, () => {
      setTimeout(() => {
        Alert.alert('Alert', 'Works fine');
      }, 510);
    });
  }
  render() {
    const { showModal } = this.state;
    return (
      <View style={styles.container}>
        <Text onPress={this.onShowModal}>Show modal</Text>
        <Modal animationType='slide' visible={showModal} onRequestClose={this.onCloseModal3} >
          <View style={styles.container}>
            <Text onPress={this.onCloseModal1}>Close modal immediately</Text>
            <Text onPress={this.onCloseModal2}>Close modal after 200ms</Text>
            <Text onPress={this.onCloseModal3}>Close modal after more then 500ms</Text>
          </View>
        </Modal>
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'space-around',
  },
});


export default demo;
```

**Comment by leeight** [comment]
+1

**Comment by alma-socar** [comment]
`InteractionManager.runAfterInteractions` also doesn't work in this case

**Comment by ganmor** [comment]
The opposite is also true, showing an alert in the same loop just before showing a modal will prevent the modal from being displayed

**Comment by nihgwu** [comment]
Yes, I guess they are the some issue, so I just make it simple to demonstrate

? 2016?10?20??20:07?Morgan Laupies <notifications@github.com<mailto:notifications@github.com>> ???


The opposite is also true, showing an alert in the same loop just before showing a modal will prevent the modal from being displayed

-

**Comment by K-Leon** [comment]
I noticed same issue with experimental navigator. An Error while transitioning kills the animation and freezes UI.

**Comment by scarlac** [comment]
Same issue here. Discovered it while during an upgrade of React from 0.34 to 0.35. Happens consistently.

**Comment by ericvicenti** [comment]
Looks like we need `<Modal>` animations to tie into `InteractionManager`. A PR would be appreciated for this!

Any issues with NavigationExperimental should be filed separately.

**Comment by nihgwu** [comment]
@ericvicenti I didn't use NE or Navigator in demo at all, and I don't think this issue is related to `InteractionManager`, my guess is that `Modal` and `Alert` are using the same controller or view, so there will be a conflict when trying to control them simultaneously, I know nothing about iOS or I would be glad to make a PR, because it's really annoying but has a quite common use case

In my demo, I've show the different results when waiting for a certain milliseconds. I guess the `Modal` need 500ms to dismiss, so it's safe to show the `Alert` after 500ms. Perhaps you are right to tie `Modal` animations into `InteractionManager`, but what if I show the `Alert` first as @ganmor mentioned above, I can't close the `Alert` manually, or if I could, would there be the same occasion that we need to wait for the `InteractionManager` before showing the `Modal`?

Would someone expert in iOS take a look at this issue?

**Comment by nihgwu** [comment]
ping @grabbou

**Comment by DevBkIL** [comment]
do someone have a fix for it?

**Comment by jasonmerino** [comment]
@DevBkIL my not great solution is to set a timeout, enough for the presented view controller to be the one under the Modal, and then present the Alert. Like I said, not great, but it's working until this can be properly resolved.

**Comment by nihgwu** [comment]
UPDATE: UI will be blocked when show Share/ActionSheet while closing Modal too
And if we are showing the Alert/Share/ActionSheet, then show a Modal, the UI will be froze too
ping @javache @mkonicek

**Comment by joegoodall1** [comment]
Any update/resolution to this? We're still having problems with this a month after this issue was first reported.

**Comment by mjamiesoncs** [comment]
+1

**Comment by sibelius** [comment]
I can confirm this using https://github.com/jaysoo/react-native-prompt

any workaround?

**Comment by baurine** [comment]
today I met the similar problem after I upgraded the react-native from 0.33 to 0.37. I want to show an Alert dialog after close the Modal, but Modal doesn't disappear, even after I close the Alert dialog and use `cmd + R` to reload the app. only in iOS, and it works fine by react-native 0.33.

the code likes following:

```
  _pressNext() {
    // display a Modal with a spinner
    this.setState({isProcessing: true}}

    // network request
    // ...
  }

  componentWillReceiveProps(nextProps) {
      // ...

      // to hide the Modal with a spinner
      this.setState({isProcessing: false})
      Alert.alert('title', 'Something has done!', [
        { text: 'Got it', onPress: () => {} }
      ])
    }
  }
```

then I try to use `setTimeout` to work around it, the code likes following: 

```
  componentWillReceiveProps(nextProps) {
      // ...

      // to hide the Modal with a spinner
      this.setState({isProcessing: false})
      setTimeout( () => {
        console.log("show alert")  // this log will ouput
        Alert.alert("title", "msg")   // but Alert doesn't display
      }, 200)
  }
```

then the Modal will disappear, but, the Alert dialog can't display!!!

I also tried run `setTimeout` in `setState` callback, like this:

```
  this.setState({isProcessing: false}, () => {
    setTimeout( () => {
      Alert.alert("title", "msg")
    }
  }
```

but the same result, Alert dialog doesn't pop up yet.

finally, I decide to hide Modal really after I close the Alert dialog, it works! code likes following:

```
Alert.alert("title", "msg", [
  { text: "OK", onPress: () => { this.setState({ isProcessing: false } }    
])
```

**Comment by vicmpen** [comment]
Had the same problem with RN 35, a little higher timeout worked for me, but ultimately i also did what you did, hide the modal on OK Pressed in Alert

`return Alert.alert('Σφάλμα','Η υπηρεσία δεν είναι διαθέσιμη αυτή τη στιγμή. Παρακαλούμε δοκιμάστε σε λίγο.',
                            [  {text: 'OK', onPress: () =>this.setState ({spinnerIsVisible:false})} ])`

**Comment by b8ne** [comment]
We are having a similar issue in [niftylettuce/react-native-loading-spinner-overlay](https://github.com/niftylettuce/react-native-loading-spinner-overlay/issues/30). Could it be that this is a more general issue relating to UI updating on state change rather than just a modal issue?

**Comment by ericvicenti** [comment]
cc @javache

**Comment by AlimovSV** [comment]
+1

**Comment by mattotodd** [comment]
+1

**Comment by susan-github** [comment]
+1,  anyone has found solutions?

**Comment by K-Leon** [comment]
@b8ne @javache It is a general error #10598 that has been introduced more recently.

**Comment by NoelBroda** [comment]
+1 Same issue here!
If I run `this.setState` to close the modal, and after that I run a `Alert.alert`, the Modal will not be closed.

**Comment by susan-github** [comment]
Hope someone can follow up this issue. Thanks!

**Comment by nihgwu** [comment]
@mkonicek what about to assign to someone? One and a half months have passed

**Comment by ericvicenti** [comment]
Can somebody help find the commit that broke this? Usually I do that by bisecting and by looking at the history of relevant files. Once we find the cause of the regression, we can fix it and/or assign the bug to whoever authored that commit.

**Comment by nihgwu** [comment]
@ericvicenti I think this regrression was introduced by https://github.com/facebook/react-native/commit/d8b2bab794841c354e70f7e707eaaea98e19e8ed so @mmmulani should be the potential assignee as he made all the changes on `Modal` in RN0.34

**Comment by susan-github** [comment]
After asked iOS developer, it is caused by presentViewController. presentView will be show after the previous one closed completely. While after compare 0.38 with 0.33 version, the Modal component start using presentViewController, which then lead to conflict with Alert.
<img width="307" alt="dingtalk20161213151617" src="https://cloud.githubusercontent.com/assets/9947305/21131126/ed83654c-c147-11e6-80c2-e43550601a13.png">

**Comment by nihgwu** [comment]
@susan-github Thank you for your answer, the as my guess. ping @ericvicenti

**Comment by susan-github** [comment]
The solution to control the sequence by iOS code instead of setTimeout by js will be more pleasure.

**Comment by nihgwu** [comment]
ping @ericvicenti

**Comment by tomykho** [comment]
ping @ericvicenti

**Comment by weiyisheng** [comment]
+1

**Comment by mmmulani** [comment]
hey, current maintainer of Modal here. The Modal-Alert combination (or really any followup interaction from Modal closing) has been pretty tough to get working right.

I'll try to spend some time on it this Friday, I think the proper solution would be to add some kind of onClose handler that let's JS do a follow up action when we can be certain that native has navigated away already.

**Comment by Jlexyc** [comment]
The same issue...

**Comment by joegoodall1** [comment]
@mmmulani Any success sorting this problem out?

**Comment by mmmulani** [comment]
hey sorry, I got caught up with some other things at work. I looked into this a bit and it seems like #10303 might actually be able to fix this. Could any of you try it and let me know? If so we can work on getting that merged

**Comment by ghuh** [comment]
This is still broken in RN v0.40.0

**Comment by niftylettuce** [comment]
please fix 👍

**Comment by niftylettuce** [comment]
For anyone interested, a simple fix is to wrap with a `setTimeout`.

```js
// MODAL STUFF HERE
setTimeout(() => {
  Alert.alert('Oops!', err.message);
}, 100);
```

**Comment by feyy** [comment]
Your guys can try setting the Modal with`animationType="none"`, so the Modal can disappear immediately.

**Comment by lodev09** [comment]
FYI, this is still broken in RN0.41. I guess the best "official" solution for this is to implement an `onHidden`. I tried [this](https://github.com/facebook/react-native/pull/10303#issuecomment-252920282) suggestion but no luck.

My solution for now is that I'm combining `setTimeout` and @feyy's `animationType="none"`. I'm currently implementing my own custom ActionSheet so I need to show another modal "immediately" after clicking an item:

```js
constructor(props) {
  super(props);

  this.state = {
    visible: false,
    animation: "fade"
  };
}

handleModalOnShow = () => {
  this.setState({
    animation: "none"
  });
}

hideModal(callback) {
  this.setState({
    visible: false,
    animation: "fade"
  }, () => {
    setTimeout(() => {
      if (callback) callback();
    }, 100);
  });
}

render() {
  return (
    <Modal
      animationType={this.state.animation}
      transparent
      visible={this.state.visible}
      onShow={this.handleModalOnShow}
    >
      ...
    </Modal>
  );
}
```

This is the closest thing I can get. Hope this helps someone until this issue is fixed :)

**Comment by lukescott** [comment]
I'm not sure if this is the same issue, but if I hide the modal while dismissing the keyboard the modal re-appears and the contents can't be interacted with.

**Comment by Burak07** [comment]
I had the same issue.

I covered my modal with view and in this view added onstartshouldsetreponder to call closeModal callback function which makes visibility state of modal false. Method was called modal was closed but UI was blocked too until clicking screen one time.

I changed my wrapper view with a TouchableOpacity and called the same method. Modal was closed and UI was not blocked. 

Both have the same scenario I used view + onstartshouldsetresponder in first one, touchableopacity and onPress in second one.

(Note I have a few buttons in my modal I used another wrapper view to wrap them and added onStartShouoldSetResponder={(evt) => { return true}}) to make touchables that are inside of view active.

**Comment by snechiporenko** [comment]
Try this:
                            ```
new Promise((r) => {
                                setTimeout(r, 1000);
                            }).then(() => {
                                this.action();
                            });
```

**Comment by Jlexyc** [comment]
For everyone who suggest to use timers:
Maybe this bullshit codding is ok for JS or Android... but for native iOS developer it looks 100% unacceptable. Better to disable modal animation at all...

**Comment by vicmpen** [comment]
@Jlexyc Take it easy. No one is forcing you to use RN. Go with Ionic, or w/e, if you don't like it or you are not satisfied by the development work being done.

**Comment by grabbou** [comment]
It's not a surprise that it is OK for JS, it's a timeout, it's Javascript,
it's async bridge. There are a lot of concepts that make this workaround a
very natural solution.

**Comment by Jlexyc** [comment]
ionic instead of RN... really? (facepalm)

**Comment by lukescott** [comment]
I solved my problem with a timeout, but it didn't work at first. One thing I noticed was I was closing my modal without dismissing the keyboard first. This is what caused the conflict. So I had to dismiss the keyboard and then dismiss the modal after a timeout.

What interesting is it was smart enough to dismiss the keyboard on it's own when the modal was dismissed, but because the modal dismissal was in close proximity to dismissing the keyboard, a timeout did not work until I did it myself and put the delay in-between.

It seems like what's happening is the keyboard dismissal causes the modal animation to get canceled and revert, even though on screen it appears to dismiss just fine. 

The timeout work around is not perfect, but it works. Ideally the modal shouldn't reappear like it does now.

**Comment by lodev09** [comment]
I just published my first RN package after few weeks of learning/using this framework -- and yes its far more better than other cross platform out there.

The package includes a custom ActionSheet that uses the modal which works well for me (disabled the default animation and created my own). If anyone interested:
https://github.com/lodev09/react-native-cell-components

**Comment by chriteixeira** [comment]
I'm experiencing the same problem. I have also tried to add the alert in the setState callback but the problem remains:

`        this.setState({
            loadingModalVisible: false,
        }, () => alert(msg));`

**Comment by yinsee** [comment]
@chriteixeira try
```
this.setState({ loadingModalVisible: false });
setTimeout(() => alert(msg), 1);
```

**Comment by lukescott** [comment]
I have done everything mentioned above. In some places it works. In others it doesn't. Right now I am trying to dismiss a modal and show an alert - The modal won't disappear and the alert show for a brief moment and disappears. I add a timeout on top of the setState callback and the modal disappears but my alert doesn't show up.

**Comment by sowdri** [comment]
Guys, just a suggestion for anyone looking for a quick fix. 

Until this issue is fixed, it's better to do the following:

- use regular screen in the place of a modal
- hide the nav bar for this screen
- use transition bottom-to-up or fade-id

I'm suggesting this because this issue started occuring starting form some version of RN, before that it was totally fine. I did try sevaral work around as well, but it's not consistent or felt a bit hacky.

**Comment by javache** [comment]
Immediately presenting an alert while the modal is dismissing will be hard to synchronize, since this is a race between the different threads React Native is operating on. I can add something to mitigate  the other scenarios though. You should then be able to use `requestAnimationFrame` to present it correctly.

**Comment by linigithub** [comment]
setTimeout method works for me. I am using three modals.

**Comment by petersuwara** [comment]
The following does not work (Using react-native-spinner) for modal render.

`} catch({error, status}) {
            this.setState({error: error});
            console.log("error " + error);
            this.setState({isLoading: false});
            requestAnimationFrame(() =>
                Alert.alert(
                    'ERROR',
                    'Unable to connect to the server, please check your username and password.',
                    [
                        {text: 'OK', onPress: () => console.log('Error')},
                    ]
                ))
        }`

**Comment by petersuwara** [comment]
SIMPLE WORKAROUND DISCOVERED

```
catch({error, status}) {
            this.setState({error: error});
            console.log("error " + error);
            requestAnimationFrame(() => {
                this.setState({isLoading: false})
                requestAnimationFrame(() => {
                    Alert.alert(
                        'ERROR',
                        'Unable to connect to the server, please check your username and password.',
                        [
                            {text: 'OK', onPress: () => console.log("Error")},
                        ])
                })
            })
```

**Comment by kholiavko-roman** [comment]
I have the same issue

**Comment by Vetalyo** [comment]
+1
Any solutions for that?

**Comment by liucr** [comment]
+1

**Comment by esganzerla** [comment]
+1

I am not facing the problem with Alert, but closing one Modal and opening other.
Using timeout solved the problem for now.

**Comment by RobPando** [comment]
+1
It does seem to work if I set the animationType to 'none' but that is just not a viable solution for me.

**Comment by davidlongest** [comment]
Instead of using `setTimeout()`, `react-native-modal` also has a `onModalHide` prop callback that can be used to show the alert after the modal has hidden. This can be a bit less fragile.

**Comment by mikaello** [comment]
@davidlongest I tried `react-native-modal`, but I'm experiencing this bug with that modul as well. See this Expo snack with a modified version of the [`react-native-modal` example](https://github.com/react-native-community/react-native-modal/tree/master/example) with alert on `onModalHide`: https://snack.expo.io/HkaZV7f_W

**Comment by alex-golub** [comment]
solution is worked for me is:
`Alert.alert('Error', 'Please check your internet connection and try again', [
                        { text: "OK", onPress: () => {
                            this.setState({loading: false});
                        }}
                    ], { cancelable: false });`

**Comment by wmonecke** [comment]
I have the same issue when I programmatically close a Modal that is a Child of another Modal.
The Animations of both just collide and the app freezes, needing a restart.

**Comment by JJMoon** [comment]
Same issue in RN v0.48.2 with "react-native-router-flux": "^4.0.0-beta.18"
The issue started suddenly after I make inheritance. I have no clue about the cause.

**Comment by baijunjie** [comment]
+1

**Comment by shengdoushi** [comment]
+1

**Comment by JCMais** [comment]
Same issue with RN v0.48

**Comment by huyhai** [comment]
+1

**Comment by donovanCarreno** [comment]
+1

putting redux action in setTimeout doesn't work for me

**Comment by agrosner** [comment]
same problem here. wrapping a component with an onHide callback based on props is not the way to go. The React Native team should build it into the native code properly such that if we want to show successive modals, there should be no need for setTimeout or workarounds described here.

**Comment by jose920405** [comment]
+1 
react-native: ` 0.47.2`

**Comment by lslv1243** [comment]
+1

**Comment by danieloprado** [comment]
+1

**Comment by TPMinan** [comment]
+1

**Comment by JulianKingman** [comment]
If the alerts are not called from the modal (like by a background process or listener), timeouts won't work, as the timing is unpredictable. I look forward to a permanent solution, but perhaps for now a javascript-based modal could be used in some cases (like this one: https://github.com/bodyflex/react-native-simple-modal).

**Comment by mikaello** [comment]
It seems that commit a389ffbd84224b583e71cf7c1468409cbc91ec8e resolves this issue. It looks like it will be shipped with RN 0.50, try it now by upgrading to 0.50 RC 1:

`$ react-native-git-upgrade v0.50.0-rc.1`

and add this to the modal component:

```js
<Modal
   ...
   onDismiss={() => alert('Hey, no freeze')}
>
   ...
</Modal>
```

**Comment by nightflash** [comment]
The best workaround is to wait for the end of the animation and then run Alert in the next "tick" using setTimeout without the second argument. You will be independent of the animation duration:
```
InteractionManager.runAfterInteractions(() => {
    setTimeout(() => {
        Alert.alert(...)
    });
});
```

**Comment by patrys** [comment]
FYI: there's `setImmediate` but it's still an ugly workaround.

**Comment by mikaello** [comment]
RN 0.50 is now released and the `onDismiss` prop is available as a fix for this issue.

**Comment by JCMais** [comment]
This does not resolve when you open multiple modals (one inside another), and hide all of them at the same time.

**Comment by wmonecke** [comment]
How is this still an issue after a YEAR?

**Comment by jorgefspereira** [comment]
+1

**Comment by Xaber20110202** [comment]
+1

**Comment by EladAvni** [comment]
+1

**Comment by jose920405** [comment]
Still happening in `react-native: 0.51.0`

**Comment by K-Leon** [comment]
@mikaello won‘t help so much because Alert Component triggers it too and is missing this feature

**Comment by wellyshen** [comment]
Does this issue be solved?

**Comment by mmmulani** [comment]
using onDismiss should fix this particular problem. if you have multiple Modals, ensure that they're all dismissed before creating an Alert.

I'm closing this because someone has confirmed that using onDismiss fixed it for them. If you have some insane other navigation hierarchy where you're still seeing this bug, please comment with an example.

**Comment by kesiena115** [comment]
Thansk @nightflash It perhaps worth mentioning the required import:

    import {  InteractionManager } from 'react-native';

**Comment by K-Leon** [comment]
@mmmulani it is not fixing this - using stock alert alone with an Navigation action is enough to trigger this issue and alert is missing an equivalent ondismiss property working the same as an modal.

**Comment by mmmulani** [comment]
hey, yeah this issue concerns solely the combination of Modal and Alert.

could you make a new issue with the Navigation issue combined with Alert problem?

feel free to propose some solutions or put up a pull request with onDismiss for Alert. This area is pretty unmaintained because we aren't using it but I would gladly review a pull request, or give advice on getting it working

**Comment by slorber** [comment]
@mmmulani `onDismiss` is not a viable option, because we don't actually want to show an alert message on modal dismiss.

For example we might have 3 distinct actions:


```js
action1 = () => {
  alert("action1performed");
  this.closeModal()
};

action2 = () => {
  alert("action2performed");
  this.closeModal()
};

action3 = () => {
  alert("action3performed");
  this.props.navigation.goBack();
};

action4 = () => {
  alert("action4performed");
};
```

As you can understand, we may have multiple actions, which should produce different alert messages. If I were to put the alert on dismiss, this would mean that first the alert will be done after, and then I won't be able to know which actions did trigger the dismiss. For sure I can store the action somewhere and then use the stored value on the dismiss callback but this means really bad spaghetti code in perspective. 

Plus, the fact that the modal will unmount can be a side-effect of an action being performed, like action3 that does not explicitly say the modal should close but simply do a backward navigation (which will unmount the modal). (probably the case mentioned by @K-Leon)

Obviously, the example above is a simplified version, and the different actions triggering alert and backward navigations may be done at different levels in the component hierarchy. I don't expect putting an alert that is not even close to the modal component view to ever trigger an app freeze.


@mmmulani I don't understand why this issue is closed at all. onDismiss is at most a workaround. And for advanced usecases it's a really bad one. There's still a bug to resolve on this. It's not normal to have different behaviors on iOS and Android regarding alerts, and not normal that intuitive code usage does lead to a totally unusable app.

Not saying it's the fault of anyone, not saying RN contributors are bad guys, and I know I can contribute and fix the bug myself. I am just saying this issue should not be closed.

**Comment by slorber** [comment]
> could you make a new issue with the Navigation issue combined with Alert problem?

@mmmulani I don't see the point, it's still the same issue, because the navigation action does trigger modal unmounting so alerting and performing navigation action is like alerting and removing modal.

**Comment by slorber** [comment]
The workaround of @nightflash does not even work for me, using this instead of normal alert crash my app now:

```js
export const safeAlert = (...args) => {
  InteractionManager.runAfterInteractions(() => {
    setTimeout(() => {
      alert(...args);
    },0);
  });
};
```

I can't use `onDismiss` easily in my app, because all modal dismiss should not be handled the same way, and are handled at different levels in component hierarchy. 

The only viable workaround for this issue so far is to not display any alert and replace them with an absolute view layer (like toasts or something), as mentioned by @wmonecke

**Comment by mmmulani** [comment]
ahhh I see, yea the only use case we've seen is someone showing an alert after dismissing the modal. Didn't realise people were doing them both at the same time.

In that case, feel free to submit a PR to fix this, whether the issue is open

**Comment by slorber** [comment]
thanks for reopening, I'll try to help if I can but I'm not a native iOS dev unfortunately, and fixing it is not a critical issue for my app (alert is a temporary solution until I implement a better feedback system myself)

**Comment by cbartondock** [comment]
Can confirm that this is still an issue, timeout solutions didn't work for me.

**Comment by cbartondock** [comment]
Possible solution using onDismiss:
`action1 = () => {
  this.setState({action1: true});
  this.closeModal()
};

action2 = () => {
this.setState({action1: true});
  alert("action2performed");
  this.closeModal()
};

action3 = () => {
this.setState({action1: true});
  alert("action3performed");
  this.props.navigation.goBack();
};
`
And then in the Modal
`
onDismiss={()=>{
if(this.state.action1){
  this.setState({action1: false})
  alert("action1performed");
}
if(this.state.action2){
  this.setState({action2: false})
  alert("action2performed");
}
if(this.state.action2){
  this.setState({action2: false})
  alert("action2performed");
}
}}
`

**Comment by rikur** [comment]
+1

**Comment by neiker** [comment]
Very frustrating :(

**Comment by vwmoose** [comment]
+1

**Comment by henrikland** [comment]
Not sure if this is related, but I have an issue where tapping a button in a Modal should open a URL in the native browser. Works fine on Android but on iOS the browser wont open unless I first close the Modal then open the browser after a >500ms timeout...

**Comment by K-Leon** [comment]
Same thing A onDismiss is your friend.

**Comment by henrikland** [comment]
Sorry, in this case I actually don't want the modal to close, I want the browser to open and the modal to stay open in the app.

**Comment by esutton** [comment]
+1 Still a problem on RN 0.53.3

react-native 0.42.3 was working fine?

````
		"react": "^16.2.0",
		"react-native": "^0.53.3",
````

**Comment by SamiChab** [comment]
Come on, please adress this issue

**Comment by joseguya** [comment]
This issue persists on react-native 0.54.1, but @baurine work-around works.

**Comment by isaachinman** [comment]
Just encountered this issue myself, but having nothing to do with alerts at all, just opening and closing a Modal in very fast succession (network request loader).

**Comment by slorber** [comment]
@isaachinman flashing a spinner to the user is not very good, you'd rather display it a little bit longer or not at all. It may resolve the modal issue. You can try https://github.com/sindresorhus/p-min-delay

**Comment by isaachinman** [comment]
@slorber Fair enough, but this should be a design decision, not forced because the underlying tech is broken.

**Comment by slorber** [comment]
@isaachinman totally agree, just giving you this as a hint to solve your problem. As you can see I advocated to keep this issue open

**Comment by huanghaiyang** [comment]
+1

**Comment by cmbasnett** [comment]
Here we go again. With nearly every module of RN I interact with, performing some (normally) trivial task exposes fundamental bugs. Then I think: "Hmm, surely this is a standard use case, there's *no way* this hasn't been solved!". Then, inevitably, I end up here, on the issue queue (the de-facto "documentation" for this library), where I see the bug was reported nearly 2 years ago, but no one seems to have the knowledge or the desire to fix it.

**Comment by pkhien95** [comment]
+1

**Comment by aligfl** [comment]
Ran into the same issue, there is absolutely no way to workaround this :( except to use a timeout, oy vey. This sucks :(

**Comment by wadim** [comment]
# A workaround based on second parameter to setState()

I couldn't use `onDismiss` because the `Modal` I was using was within a third-party library so I had to find a different workaround.

The workaround relies on the second parameter to [`setState`](https://reactjs.org/docs/react-component.html#setstate):

> The second parameter to setState() is an optional callback function that will be executed once setState is completed and the **component is re-rendered**.

This feature allows us to hook in after the Modal has been unmounted.

## Howto

1. Add a `isModalDisabled` to component state.

2. Render the component conditionally based on whether `isModalDisabled` is true:

```js
render() {
  return (
    !this.state.isModalDisabled && <ComponentContainingAModal>
     //...
    </ComponentContainingAModal>
  );
}
```

3. When your Modal-enabled component calls a callback with the data it was supposed to retrieve do this:

```js
callbackFromMyModalEnabledComponent(someJuicyData) {
  this.setState({
    isModalDisabled: true, // This will effectively unmount the Modal (at some point)
  }, () => {
    // Here the Modal has already been unmounted so you can safely show your alert
    Alert.alert(
      'Alert!',
      'Yes or No?',
      [
        {
          onPress: () => console.log('Cancel Pressed'),
          style: 'cancel',
        },
        { onPress: () => console.log('OK Pressed') },
      ],
    );
    // Here you can set the modal to enabled again
    this.setState({
      isModalDisabled: false,
    });
  });
}
```

**Comment by jjercx** [comment]
+1 yep

**Comment by stevanus1997** [comment]
+1 any one solve this problem?

**Comment by remeryAGS** [comment]
+1, still an issue for me

**Comment by mromarreyes** [comment]
+1, still an issue for me

**Comment by trev91** [comment]
+1 still an issue

**Comment by rayj10** [comment]
+1 still an issue for me on react native v0.55.2, UI just locks, everything works fine on Android.

**Comment by d0z0** [comment]
Just to add to that, `onDismiss` is only supported on iOS, so as ugly at it may seem, if you want to display an Alert after dismissing a modal, supported on both platforms, here goes..

```
this.setState({modal_visible: false}); 
if(Platform.OS === 'android') { 
  Alert.alert('Modal dismissed');
}
```

```
<Modal onDismiss={() => Alert.alert('Modal dismissed')}/>
```

**Comment by Darking360** [comment]
The timeouts solutions have worked for me. My problem is that when the component is updated, if I was showing an Alert.alert in iOs, the alert disappears right after the setState or Redux update. I think it must have to be with some lifecycle event issue or re-rendering the component, which makes the alert in iOs to go away. I hope this to get solved in the near future. Thank you.

**Comment by edgarbonillag** [comment]
Hi Everyone! 

I want to share my SOLUTION to an specific task i needed to do. I hope it will be useful for you all.
I had two Time Pickers, one to set the opening time of a place and the other to set the closing time. The user needs to see the following:

1. When a button is pressed, an Alert is shown indicating the user to set the opening time.
2. When the Ok button on the Alert is pressed, the Alert disappears and the first Time Picker is shown.
3. When the user presses on 'Confirm', the Time Picker disappears and another Alert is shown indicating the user to set the closing time.
4. When the Ok button on the Alert is pressed, the Alert disappears and the second Time Picker is shown.
5. When the user presses on 'Confirm', the Time Picker disappears.

I experienced most of the issues exposed on this thread. And i had to combine several workarounds to make my code work.  I alse read something that suggests that Alerts and Modals (Time Picker) are ANIMATED on iOS, and the issue happens when two animations overlap. For example, in my case when going from step 3 to 4. 

So, my overall solution was to ensure that the animations of the Alerts and Time Pickers don't overlap.

I ended up doing this:

```
onOpeningTimeSet = (time) => {		
	this.setState( prevState => {
		return {
			...prevState,
			timePicker: {
				...prevState.timePicker,
				openingTimePickerVisible: false,
			}
		}
	}
	,
		() => setTimeout( () => {			
			Alert.alert(
				"",
				"Please set the CLOSING TIME",
				[	{text: "OK", onPress: () => {
						setTimeout(() => this.setEndTimePickerVisible(true), 600)
					}}
				]
			)
		}, 600)
	)
}
```

'onOpeningTimeSet' is a function that is called when the user presses the 'Confirm' button of the first Time Picker (Step 3). Inside that function, i update the state in order to hide the first Time Picker. But, as 'setState' is asynchronous, i have to make sure to show the Alert when the state is really updated and the Time Picker hides. That is easy to do by putting that code as the callback function of 'setState'. 

But, it doesn't guarantee that the will be executed after the Time Picker hiding animation is really finished. So, as you can see, i make the Alert appear 600ms after the state was updated, giving enough time to the hiding animation to finish (times below 500ms did not to work).

And then, when the user presses the Ok button of the Alert, the Time Picker will appear 600ms after, giving enough time to the Alert's hiding animation to finish.

This way, the animations won't overlap and my app runs perfectly. 

P.S: That `setEndTimePickerVisible(true)` function is just this:

```
setEndTimePickerVisible = (value) => {
	this.setState(prevState => {
		return {
			...prevState,
			timePicker: {
				...prevState.timePicker,
				closingTimePickerVisible: value
			}
		}
	})
}
```

**Comment by tiendn** [comment]
I think use **setTimeout()** is the best solution right now.
```javascript
setTimeout(() => {
      this.onReset();
      Alert.alert("Confirm failed !", "The wrong answer.");
}, 300);
```

**Comment by luisehk** [comment]
Any luck yet? I don't want to fill my code with setTimeout.

**Comment by marcshilling** [comment]
Just an FYI - while it's not ideal and should be fixed, I've had the 10ms setTimeout out in production for months now without issue.

**Comment by mujavidb** [comment]
To add to the above, it really depends on the device you are using. For lower-end devices that run slower, you need a higher timeout. My usual timeout is 300ms.

Given this is such a major issue and such a common use-case I'm still surprised it has persisted for 2 years. Any updates @ericvicenti ?

**Comment by doublex** [comment]
The 'timeout()' workaround works

**Comment by bars0um** [comment]
hmmm...1000 ms and still same issue. I'm using redux with a fetch sub-state that indicates when a modal should should. I basically reset that right before I set a timer to show my alerts within 1000 ms and still no cigar...

**Comment by Ashoat** [comment]
`Modal` isn't really being actively maintained, and is going to be moved out of the React Native core in the future as the "slimmening" commences. I'd recommend migrating off of it in favor of something like React Navigation's StackNavigator. Example:

```js
const RootNavigator = createStackNavigator(
  {
    [AppRouteName]: { screen: ReduxWrappedAppNavigator },
    [ThreadPickerModalRouteName]: { screen: ThreadPickerModal },
    [AddUsersModalRouteName]: { screen: AddUsersModal },
    [CustomServerModalRouteName]: { screen: CustomServerModal },
    [ColorPickerModalRouteName]: { screen: ColorPickerModal },
    [ComposeSubthreadModalRouteName]: { screen: ComposeSubthreadModal },
  },
  {
    headerMode: 'none',
    mode: 'modal',
    transparentCard: true,
  },
);
```

**Comment by prateekgoel** [comment]
Any progress here? This is becoming a bottleneck.

**Comment by LRNZ09** [comment]
I'm also having this issue on RN `0.57.8`

**Comment by ptupi** [comment]
+1

**Comment by gsandiford** [comment]
+1

**Comment by dev-galactica** [comment]
> `Modal` isn't really being actively maintained, and is going to be moved out of the React Native core in the future as the "slimmening" commences. I'd recommend migrating off of it in favor of something like React Navigation's StackNavigator. Example:
> 
> ```js
> const RootNavigator = createStackNavigator(
>   {
>     App,
>     ThreadPickerModal,
>     AddUsersModal,
>     CustomServerModal,
>     ColorPickerModal,
>     ComposeSubthreadModal,
>   },
>   {
>     headerMode: 'none',
>     mode: 'modal',
>     transparentCard: true,
>   },
> );
> ```

In some cases, it is really useful to have the Modal component available. I do not consider the stack navigator with modal transition as a complete replacement for it.

I work with React Navigation Stack Navigator, however, we need some alternative flows for tablets for our custom flow (which is similar to iOS SplitView design to make advantage of tablets size), in those flows, Modal becomes handy when we have to display something in top of the rest of the whole UI.

An example:
Left pane: Stack Navigator (1), Right pane: Stack Navigator (2)
...
In top of it: Modal

For this scenario, if you use Stack Navigator modal transition, you cannot display something in the whole screen. You can however, display that in the left of right pane.

**Comment by Ashoat** [comment]
You can always just create a `StackNavigator` that wraps the rest of your navigators and use that for displaying modals.

**Comment by dev-galactica** [comment]
> You can always just create a `StackNavigator` that wraps the rest of your navigators and use that for displaying modals.

I see your point. It may help without doubt, it seems without Modal it would be the only direct option.
I think the cluttering in the 'parent' navigator can be important in case we do that, because indirect child navigators would use that ancestor navigator to replace the UI.

About the 'modal' option in Stack Navigators, I appreciate the inconveniences below:
- What happens if you have to open a modal upon another? We do that with Modal component without problems, e.g. show a list of elements in a modal, which display a full screen image when pressed.
- The modal in React Navigation seems to me like a different transition, but it does not block the components below it. So you are not really opening a modal as in Native apps, you are tricking the inner Activity or ViewController (it is fine if you achieve same results anyway).
- The Stack Navigator modals can be dismissed on iOS if the user swipes from top to bottom, it is sometimes good, but others can result in the user leaving the screen by error.

**Comment by Jlexyc** [comment]
I've build my own modal with `position='absolute'` and it works perfectly without additional delays for Alert.
Also you can build custom appearance animation for it. So just don't use modal out of the box.

**Comment by webdevbyjoss** [comment]
Hey guys, have the same issue right now. Are there any plans to resolve this in near future?

**Comment by quynhnguyen68** [comment]
Hi everyone, I solved this problem by adding 100ms timeout as recommended in the comments above. I guess we need to set timeout and waiting for the modal close first
```
    setTimeout(() => {
      Alert.alert(name,  message);
    }, 100);
```

**Comment by cristianoccazinsp** [comment]
Was this ever fixed? Do we really need to resort to the good ol' timeout "fix"?

**Comment by bars0um** [comment]
If you are using react-navigation and the app reloads, each page upto the one you were on will mount. If you have a modal being shown on two of those screens at mounting time (componentDidMount), they'll clobber each other...

**Comment by housenkui** [comment]
At iOS side,I do this work well.

UIViewController *presentedViewController = [UIApplication sharedApplication].keyWindow.rootViewController.presentedViewController;
    NSOperationQueue *mainQueue = [NSOperationQueue mainQueue];
    
    mainQueue.maxConcurrentOperationCount = 1;
    
    NSBlockOperation *dismissOperation = [NSBlockOperation blockOperationWithBlock:^{
        if (presentedViewController) {
            [presentedViewController dismissViewControllerAnimated:NO completion:^{}];
        }
    }];
    NSBlockOperation *presentOperation = [NSBlockOperation blockOperationWithBlock:^{
        [rootVC presentViewController:scanIdVC animated:YES completion:nil];
    }];
    [presentOperation addDependency:dismissOperation]; // 让presentOperation 依赖于 dismissOperation，则先执行dismissOperation，再执行presentOperation

    [mainQueue addOperation:dismissOperation];
    [mainQueue addOperation:presentOperation];

**Comment by housenkui** [comment]
`UIViewController *presentedViewController = [UIApplication sharedApplication].keyWindow.rootViewController.presentedViewController;
NSOperationQueue *mainQueue = [NSOperationQueue mainQueue];

mainQueue.maxConcurrentOperationCount = 1;

NSBlockOperation *dismissOperation = [NSBlockOperation blockOperationWithBlock:^{
    if (presentedViewController) {
        [presentedViewController dismissViewControllerAnimated:NO completion:^{}];
    }
}];
NSBlockOperation *presentOperation = [NSBlockOperation blockOperationWithBlock:^{
    [rootVC presentViewController:scanIdVC animated:YES completion:nil];
}];
[presentOperation addDependency:dismissOperation]; // 让presentOperation 依赖于 dismissOperation，则先执行dismissOperation，再执行presentOperation

[mainQueue addOperation:dismissOperation];
[mainQueue addOperation:presentOperation];`

**Comment by housenkui** [comment]
At iOS side,I do this work well. I hope the code can help you.


```
 UIViewController *presentedViewController = [UIApplication sharedApplication].keyWindow.rootViewController.presentedViewController;
    NSOperationQueue *mainQueue = [NSOperationQueue mainQueue];
    
    mainQueue.maxConcurrentOperationCount = 1;
    
    NSBlockOperation *dismissOperation = [NSBlockOperation blockOperationWithBlock:^{
        if (presentedViewController) {
            [presentedViewController dismissViewControllerAnimated:NO completion:^{}];
        }
    }];
    NSBlockOperation *presentOperation = [NSBlockOperation blockOperationWithBlock:^{
        [rootVC presentViewController:scanIdVC animated:YES completion:nil];
    }];
    [presentOperation addDependency:dismissOperation]; // 让presentOperation 依赖于 dismissOperation，则先执行dismissOperation，再执行presentOperation

    [mainQueue addOperation:dismissOperation];
    [mainQueue addOperation:presentOperation];
```

**Comment by xaamin** [comment]
In which file you made this change @housenkui ? Thanks in advance.

**Comment by cristianoccazinsp** [comment]
Any updates on this? Still happening...

**Comment by smilecc** [comment]
same question!
It's very incredible. I found, the question always appears in the real machine, but not in the simulator.

**Comment by LRNZ09** [comment]
@housenkui can you tell which file is it?

**Comment by rogerbright** [comment]
The timeout workaround seems to work but it's clunky at best. There's no guarantee that 500ms (or any arbitrary time) will always be enough. Are the developers too busy making self-indulgent breaking changes to all the other RN APIs to fix basic issues? (Sorry, I'm frustrated.)

**Comment by cltsang** [comment]
If we are using navigation libraries from third parties anyway, you should use modal from third party library too, as one could argue that it belongs to navigation.

If you use react-navigation, just use StackNavigator with transparentCard set to true, and mode set to 'modal'.
See: https://reactnavigation.org/docs/en/stack-navigator.html#stacknavigatorconfig

In my opinion, navigation is such a fundamental interaction that it should be included in the core. But it is what it is.

**Comment by zhongwuzw** [comment]
Hi all, I'll be closing this because we may fixes it in master branch, feel free to reopen if you think bug not fixed.

**Comment by IlarionHalushka** [comment]
still have it in rn 0.59.8

**Comment by zhongwuzw** [comment]
> still have it in rn 0.59.8

@IlarionHalushka Hey, we still don't release yet, please test on `master` branch.

**Comment by IlarionHalushka** [comment]
oh, sorry for my misunderstanding... Unfortunately I wasn't able to test on `master` branch because ios build failed for me:(
For now I fixed the issue with work-around by hiding modal on Alert OK press. Anyway thanks a lot!

**Comment by maxKimoby** [comment]
2 years alter and it is still a thing. Can't hide 2 modals in a single setState call because of this. You got to manually set a timeout for the second one to close.

**Comment by deepasurya1c** [comment]
Still having this issue. I am showing a loader on screen which is a Modal and then I need to show popup after API response. It works on Android to hide loader on Ok button action of Alert popup, but not working on iOS. Loader gets stuck on screen and popup never appears. Need to kill the app and restart it. Can anyone have solution?

**Comment by deepasurya1c** [comment]
> oh, sorry for my misunderstanding... Unfortunately I wasn't able to test on `master` branch because ios build failed for me:(
> For now I fixed the issue with work-around by hiding modal on Alert OK press. Anyway thanks a lot!

It worked on android but not on iOS. Please suggest some solution.

**Comment by zhongwuzw** [comment]
Hi guys, I think we may fixed it on latest `master` branch, if you guys tried it which not worked, please file a new issue.

**Comment by TechSatya** [comment]
Using of "AlertIOS" component for iOS helped me to avoid this issue.

The workaround is as below;

```
AlertIOS.alert(
  'Alert Title', 'Alert message',
    [
      {
        text: 'OK', onPress: () => this.setState({ loading: false})
      }
    ]
);
```

**Comment by cristianoccazinsp** [comment]
@TechSatya  are you confident about this? One would guess that Alert uses AlertIOS behind the scenes.

**Comment by TechSatya** [comment]
Yes, @cristianoccazinsp I think this is the only possible solution as of now because the other availed solution didn't help me in any way. As from the last 2 days I was stuck in this issue, I started giving a try to the "AlertIOS" component and fortunately found it successful.

Yes, Alerts uses AlertIOS behind it, but don't know how it got successful? Maybe something lacks in implementing of AlertIOS in Alert behind the scenes.

To be more specific to the code for both platforms;
````
(Platform.OS === 'android' ? Alert : AlertIOS).alert(
  'Alert Title', 'Alert message',
    [
      {
        text: 'OK', onPress: () => this.setState({ loading: false})
      }
    ]
);
````

**Comment by fengyouchao** [comment]
Using `setTimeout` is **NOT** a safe way to avoid this problem. I recommend using `onDismiss` props in iOS to do something after the modal closed (including `Alert` some messages). I created a class named `ModalView` to replace RN `Modal` class.

```javascript
class ModalView extends React.Component {

    state = {
        visible: false
    };

    close({then} = {}) {
        if (Platform.OS === 'ios') {
            this.setState({visible: false, onDismiss: then});
        } else {
            this.setState({visible: false});
            if (then !== undefined) {
                then();
            }
        }
    }

    show() {
        this.setState({visible: true});
    }

    render() {
        return (
            <Modal
                visible={this.state.visible}
                onDismiss={this.state.onDismiss}
                {...this.props}>
                {this.props.children}
            </Modal>
        )
    }
}
```

# Using

Just replace your `Modal` with `ModalView` and add `ref` props.
```javascript
<ModalView ref={ref=>this.modalView} 
   //other Modal props
>
//your components
<ModalView>
```
## Show Modal

```javascript
this.modalView.show();
```

## Close Modal
```javascript
this.modalView.close();
```

Alert some messages after the modal closed.
```javascript
this.modalView.close({
      then: () => {
          Alert.alert('Hello World');
          //your code
      }
});
```

**Comment by 1612844** [comment]
I have same issue here when i use RN 0.59.9

And my solution is show the alert in modal onDismiss callback

**Comment by jvidalv** [comment]
The issue is still going with "**react**" : "16.8.3" and  "**react-native**" : sdk-33.0.0  using "**expo**": "^33.0.0"

The workaround is simple, yet a bit ugly, just use setTimeout() to trigger the alert:

`setTimeout(() => Alert.alert('With just 10 miliseconds already works!'), 10)`

**Comment by giautm** [comment]
Agree with @fengyouchao, your solution was working well. Thank you.

**Comment by hoanglm4** [comment]
I think facebook engineering need fix this bug.

**Comment by adimshev** [comment]
You can use my solution (prevent opening Modal when Alert is opened, and open it after Alert was closed, using redux and hooks). 

[Modal + Alert.zip](https://github.com/facebook/react-native/files/3937150/Modal.%2B.Alert.zip)

**Comment by teledemic** [comment]
FWIW a modification to @fengyouchao's workaround if you are using async/await is 
```
async Close() {
	return new Promise((resolve, _reject) => {
		this.setState({ visible: false, onDismiss: resolve });
	});
}
```

Then you can just `await this.modalView.Close();`

**Comment by jbenzshawel** [comment]
Still an issue with RN 0.61.2 - does anyone if this is fixed in any later 0.61 patch releases?

**Comment by yogesh1490** [comment]
> When show Alert while closing Modal, the Alert dialog will disappear and the Modal will block the UI entirely even after reload, only on iOS.
> 
> ```js
> 'use strict';
> 
> import React, { Component } from 'react';
> 
> import {
>   StyleSheet,
>   View,
>   Text,
>   Modal,
>   Alert,
> } from 'react-native';
> 
> class demo extends Component {
>   state = {
>     showModal: false,
>   }
> 
>   onShowModal = () => {
>     this.setState({ showModal: true });
>   }
>   onCloseModal1 = () => {
>     this.setState({ showModal: false }, () => {
>       Alert.alert('Alert', 'UI will be blocked by the modal');
>     });
>   }
>   onCloseModal2 = () => {
>     this.setState({ showModal: false }, () => {
>       setTimeout(() => {
>         Alert.alert('Alert', 'Alert won\'t show');
>       }, 200);
>     });
>   }
>   onCloseModal3 = () => {
>     this.setState({ showModal: false }, () => {
>       setTimeout(() => {
>         Alert.alert('Alert', 'Works fine');
>       }, 510);
>     });
>   }
>   render() {
>     const { showModal } = this.state;
>     return (
>       <View style={styles.container}>
>         <Text onPress={this.onShowModal}>Show modal</Text>
>         <Modal animationType='slide' visible={showModal} onRequestClose={this.onCloseModal3} >
>           <View style={styles.container}>
>             <Text onPress={this.onCloseModal1}>Close modal immediately</Text>
>             <Text onPress={this.onCloseModal2}>Close modal after 200ms</Text>
>             <Text onPress={this.onCloseModal3}>Close modal after more then 500ms</Text>
>           </View>
>         </Modal>
>       </View>
>     );
>   }
> }
> 
> const styles = StyleSheet.create({
>   container: {
>     flex: 1,
>     alignItems: 'center',
>     justifyContent: 'space-around',
>   },
> });
> 
> 
> export default demo;
> ```
I have same issue.
Has anyone done this using hooks  as useState in functional component ?

**Comment by revich2** [comment]
I have a similar problem, but the alert may appear at an undetermined point in time and is not logically dependent on Modal component.

My workaround looks like this:

``` js
  const OldAlert = Alert.alert;

  Alert.alert = (...args) => {
    requestAnimationFrame(() => {
      requestAnimationFrame(() => {
        OldAlert(...args);
      });
    });
  };
```

This is “magic,” but it works for me.

**Comment by cristianoccazinsp** [comment]
@revich2  interesting... Why do you need 2 `requestAnimationFrame` ? Would it be the same with `runAfterInteractions` ?

@react-native-bot  can we re open this?

**Comment by revich2** [comment]
@cristianoccazinsp The first thing I used was `InteractionManager.runAfterInteractions`, but for some reason, this did not work for me, so I had to make 2 `requestAnimationFrame` to defer the call to `Alert.alert` to the next iteration in the event loop.

**Comment by phamlap228** [comment]
in my case im change others Modal and resolved :D

**Comment by thanhluantl2304** [comment]
I still face that problem when close one modal and open other modal immediately. My current solution is using setTimeout but that's inconvenient. Have Facebook staff fixed it yet? Or someone has another solution?

**Comment by mangBert** [comment]
Same here. any update on this?

**Comment by scarlac** [comment]
**Cause:**
Rendering a modal is a polite request. It does not guarantee it will show. The root cause is that the native iOS implementation assumes it's a guarantee.
This is a native bug that needs to be fixed in the Modal component on iOS. The root cause is that modals are imperative in iOS and you may _request_ to open one, but it may not succeed. iOS implementation does not consider that the modal potentially never opens, so it still triggers the underlying lock-up mechanism. This leaves the app half locked up with no modal.

**Related bugs:**
This bug will also happen if you [try to use `pageSheet` presentation style](https://github.com/facebook/react-native/issues/26892) and user swipes down on the sheet. This user action causes the modal to disappear, yet React Native thinks it's still visible, leaving your app in a locked state. Same cause.

**Workaround:**
If you are okay with the modal not showing, but just want to not lock up your app, wrap your `<Modal>` in a View with 0 height, like this:
```jsx
function BrokenModal() {
  alert('this alert will show, modal wont');
  return (
    <View style={{height: 0}}>
      <Modal animated>
        <Text>This text will never appear</Text>
      </Modal>
    </View>
  );
}
```
